presto-main/src/main/java/com/facebook/presto/operator/scalar/StringFunctions.java

```diff
@@ -89,23 +90,81 @@ public static Slice replace(@SqlType(StandardTypes.VARCHAR) Slice str, @SqlType(
-                search.toString(UTF_8),
-                replace.toString(UTF_8));
-        return Slices.copiedBuffer(replaced, UTF_8);
+        //
+        // Empty search returns original string
+        if (UnicodeUtil.isEmpty(search)) {
+            return string;
+        }
+        //
+        // Allocate a reasonable buffer
+        Slice buffer = Slices.allocate(string.length());
+
+        int index = 0;
+        int indexBuffer = 0;
+        while (index < string.length()) {
+            final int matchIndex = UnicodeUtil.findUtf8IndexOfString(string, index, string.length(), search);
+            //
+            // Found a match?
+            if (matchIndex < 0) {
+                //
+                // No match found so copy the rest of string
+                final int bytesToCopy = string.length() - index;
+                buffer = Slices.ensureSize(buffer, indexBuffer + bytesToCopy);
+                buffer.setBytes(indexBuffer, string, index, bytesToCopy);
+                indexBuffer += bytesToCopy;
+
+                break;
+            }
+
+            final int bytesToCopy = matchIndex - index;
+            buffer = Slices.ensureSize(buffer, indexBuffer + bytesToCopy + replace.length());
+            //
+            // Non empty match?
+            if (bytesToCopy > 0) {
+                buffer.setBytes(indexBuffer, string, index, bytesToCopy);
+                indexBuffer += bytesToCopy;
+            }
+            //
+            // Non empty replace?
+            if (!UnicodeUtil.isEmpty(replace)) {
+                buffer.setBytes(indexBuffer, replace);
+                indexBuffer += replace.length();
+            }
+            //
+            // Continue searching after match
+            index = matchIndex + search.length();
+        }
+
+        return Slices.copyOf(buffer, 0, indexBuffer);
     }
 
     @Description("reverses the given string")
     @ScalarFunction
     @SqlType(StandardTypes.VARCHAR)
-    public static Slice reverse(@SqlType(StandardTypes.VARCHAR) Slice slice)
+    public static Slice reverse(@SqlType(StandardTypes.VARCHAR) Slice string)
     {
-        Slice reverse = Slices.allocate(slice.length());
-        for (int i = 0, j = slice.length() - 1; i < slice.length(); i++, j--) {
-            reverse.setByte(j, slice.getByte(i));
+        final Slice reverse = Slices.allocate(string.length());
+
+        for (int i = 0, j = string.length(); i < string.length(); ) {
+            final int startByte = string.getUnsignedByte(i);
+            final int codePointLength = UnicodeUtil.lengthOfCodePoint(startByte);
+
+            if (codePointLength == 1) {
+                // In the special case of length 1, we can set the byte directly
+                j--;
+                reverse.setByte(j, startByte);
+            }
+            else {
+                j -= codePointLength;
+                // TODO Illegal start bytes could end in IOOB. Discuss this in review if it is ok.
```

Comment:
Turn it into `new PrestoException(INVALID_FUNCTION_ARGUMENT, "a human readable string description", e)`